executor/index_lookup_merge_join.go

```diff
@@ -0,0 +1,673 @@
+	}
+
+	task.memTracker.Consume(task.outerResult.MemoryUsage())
+	oldMemUsage := task.outerResult.MemoryUsage()
+	err := Next(ctx, omw.executor, task.outerResult)
+	if err != nil {
+		return task, err
+	}
+
+	newMemUsage := task.outerResult.MemoryUsage()
+	task.memTracker.Consume(newMemUsage - oldMemUsage)
+	if task.outerResult == nil || task.outerResult.NumRows() == 0 {
+		return nil, nil
+	}
+	if omw.outerMergeCtx.needOuterSort {
+		c := newFirstChunk(omw.executor)
+		var rows []chunk.Row
+		for i := 0; i < task.outerResult.NumRows(); i++ {
+			rows = append(rows, task.outerResult.GetRow(i))
+		}
+		sort.Slice(rows, func(i, j int) bool {
+			for id, joinKey := range omw.joinKeys {
+				cmp, _, err := omw.compareFuncs[id](omw.ctx, joinKey, joinKey, rows[i], rows[j])
+				terror.Log(err)
+				if cmp != 0 || omw.nextColCompareFilters == nil {
+					return cmp < 0
+				}
+				return omw.nextColCompareFilters.CompareRow(rows[i], rows[j]) < 0
+			}
+			return false
+		})
+		for i := 0; i < len(rows); i++ {
+			c.AppendRow(rows[i])
+		}
+		task.outerResult = c
+	}
+
+	if omw.filter != nil {
+		var err error
+		task.outerMatch, err = expression.VectorizedFilter(omw.ctx, omw.filter, chunk.NewIterator4Chunk(task.outerResult), task.outerMatch)
+		if err != nil {
+			return task, err
+		}
+		task.memTracker.Consume(int64(cap(task.outerMatch)))
+	}
+	task.outerIter = chunk.NewIterator4Chunk(task.outerResult)
+	return task, nil
+}
+
+func (omw *outerMergeWorker) increaseBatchSize() {
+	if omw.batchSize < omw.maxBatchSize {
+		omw.batchSize *= 2
+	}
+	if omw.batchSize > omw.maxBatchSize {
+		omw.batchSize = omw.maxBatchSize
+	}
+}
+
+func (imw *innerMergeWorker) run(ctx context.Context, wg *sync.WaitGroup, cancelFunc context.CancelFunc) {
+	var task *lookUpMergeJoinTask
+	defer func() {
+		wg.Done()
+		if r := recover(); r != nil {
+			cancelFunc()
+		}
+	}()
+
+	for ok := true; ok; {
+		select {
+		case task, ok = <-imw.taskCh:
+			if !ok {
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+
+		err := imw.handleTask(ctx, task)
+		task.doneErr <- err
+		close(task.doneErr)
+	}
+}
+
+func (imw *innerMergeWorker) handleTask(ctx context.Context, task *lookUpMergeJoinTask) error {
+	dLookUpKeys, err := imw.constructDatumLookupKeys(task)
+	if err != nil {
+		return err
+	}
+	dLookUpKeys = imw.dedupDatumLookUpKeys(dLookUpKeys)
+	imw.innerExec, err = imw.readerBuilder.buildExecutorForIndexJoin(ctx, dLookUpKeys, imw.indexRanges, imw.keyOff2IdxOff, imw.nextColCompareFilters)
+	if err != nil {
+		return err
+	}
+	defer terror.Call(imw.innerExec.Close)
+	_, err = imw.fetchNextInnerResult(ctx, task)
+	if err != nil {
+		return err
+	}
+	err = imw.doMergeJoin(ctx, task)
+	return err
+}
+
+func (imw *innerMergeWorker) fetchNewChunkWhenFull(ctx context.Context, task *lookUpMergeJoinTask, chk **chunk.Chunk) bool {
+	if !(*chk).IsFull() {
+		return true
+	}
+	select {
+	case task.results <- &indexMergeJoinResult{*chk, imw.joinChkResourceCh}:
```

Comment:
If it is not the current task, the selection here is always blocked, so this task cannot get next chunk from `joinChkResourceCh` to continue its job.
This will cause that each `innerMegerWorker`  can only get one chunk finished before the task becomes to the current task.

Reply:
That the core to keep outer rows order...

Reply:
My fault, I forget `task.results` has a 4 sized buffer.